File: macos/Sources/xml-macker/ShareTextPolicy.swift
import Foundation

/// A provider-neutral policy for moving source text into another app.
///
/// Chat web sites and macOS sharing extensions do not expose one common input
/// limit. xml-macker therefore offers a conservative excerpt for predictable
/// sharing, while leaving the full selection available as an explicit copy.
/// Excerpts always end between Swift `Character`s, so an extended grapheme
/// cluster (for example an emoji family or a letter plus combining marks) is
/// never split.
struct ShareTextExcerpt: Equatable {
    let text: String
    let totalUTF8Bytes: Int
    let includedUTF8Bytes: Int

    var omittedUTF8Bytes: Int { totalUTF8Bytes - includedUTF8Bytes }
    var isTruncated: Bool { omittedUTF8Bytes > 0 }
}

enum ShareTextPolicy {
    /// Roughly 32k tokens for ordinary text. This is deliberately described as
    /// a compatibility excerpt, not a limit guaranteed by any chat provider.
    static let compatibilityUTF8ByteLimit = 128 * 1024

    static func excerpt(
        from source: String,
        maximumUTF8Bytes: Int = compatibilityUTF8ByteLimit
    ) -> ShareTextExcerpt {
        let limit = max(0, maximumUTF8Bytes)
        let totalBytes = source.utf8.count
        guard totalBytes > limit else {
            return ShareTextExcerpt(
                text: source,
                totalUTF8Bytes: totalBytes,
                includedUTF8Bytes: totalBytes
            )
        }

        var end = source.startIndex
        var includedBytes = 0
        while end < source.endIndex {
            let next = source.index(after: end) // advances by one Character
            let characterBytes = source[end..<next].utf8.count
            guard includedBytes + characterBytes <= limit else { break }
            includedBytes += characterBytes
            end = next
        }

        return ShareTextExcerpt(
            text: String(source[..<end]),
            totalUTF8Bytes: totalBytes,
            includedUTF8Bytes: includedBytes
        )
    }
}

File: windows/src/xml-macker/Shared/ShareTextPolicy.cs
using System;
using System.Globalization;
using System.Text;

namespace XMLMacker.Shared;

public readonly record struct ShareTextExcerpt(string Text, int TotalUtf8Bytes, int IncludedUtf8Bytes)
{
    public int OmittedUtf8Bytes => TotalUtf8Bytes - IncludedUtf8Bytes;
    public bool IsTruncated => OmittedUtf8Bytes > 0;
}

public static class ShareTextPolicy
{
    public const int CompatibilityUtf8ByteLimit = 128 * 1024;

    public static ShareTextExcerpt Excerpt(string source, int maximumUtf8Bytes = CompatibilityUtf8ByteLimit)
    {
        int limit = Math.Max(0, maximumUtf8Bytes);
        int total = Encoding.UTF8.GetByteCount(source);
        if (total <= limit) return new(source, total, total);

        var enumerator = StringInfo.GetTextElementEnumerator(source);
        int chars = 0;
        int bytes = 0;
        while (enumerator.MoveNext())
        {
            string element = enumerator.GetTextElement();
            int elementBytes = Encoding.UTF8.GetByteCount(element);
            if (bytes + elementBytes > limit) break;
            bytes += elementBytes;
            chars = enumerator.ElementIndex + element.Length;
        }
        return new(source[..chars], total, bytes);
    }
}
